File: titanium-rdf-api/src/main/scala/eu/ostrzyciel/jelly/convert/titanium/internal/TitaniumDecoderConverter.scala
package eu.ostrzyciel.jelly.convert.titanium.internal

import eu.ostrzyciel.jelly.convert.titanium.internal.TitaniumRdf.*
import eu.ostrzyciel.jelly.core.ProtoDecoderConverter

/**
 * A Jelly decoder converter for the titanium-rdf-api.
 */
private[titanium] final class TitaniumDecoderConverter extends ProtoDecoderConverter[Node, String, Quad, Quad]:
  override def makeSimpleLiteral(lex: String): Node = SimpleLiteral(lex)
  override def makeLangLiteral(lex: String, lang: String): Node = LangLiteral(lex, lang)
  override def makeDtLiteral(lex: String, dt: String): Node = DtLiteral(lex, dt)
  override def makeDatatype(dt: String): String = dt
  override def makeBlankNode(label: String): Node = "_:".concat(label)
  override def makeIriNode(iri: String): Node = iri
  override def makeTripleNode(s: Node, p: Node, o: Node): Node =
    throw new NotImplementedError("The titanium-rdf-api implementation of Jelly does not support " +
      "quoted triples.")
  override def makeDefaultGraphNode(): Node = null

  override def makeTriple(s: Node, p: Node, o: Node): Quad = Quad(
    s.asInstanceOf[String],
    p.asInstanceOf[String],
    o,
    null
  )

  override def makeQuad(s: Node, p: Node, o: Node, g: Node): Quad = Quad(
    s.asInstanceOf[String],
    p.asInstanceOf[String],
    o,
    g.asInstanceOf[String]
  )

File: titanium-rdf-api/src/main/java/eu/neverblink/jelly/convert/titanium/internal/TitaniumDecoderConverter.java
package eu.neverblink.jelly.convert.titanium.internal;

import eu.neverblink.jelly.core.InternalApi;
import eu.neverblink.jelly.core.ProtoDecoderConverter;

/**
 * A Jelly decoder converter for the titanium-rdf-api.
 */
@InternalApi
public final class TitaniumDecoderConverter implements ProtoDecoderConverter<Object, String> {

    @Override
    public Object makeSimpleLiteral(String lex) {
        return new TitaniumLiteral.SimpleLiteral(lex);
    }

    @Override
    public Object makeLangLiteral(String lex, String lang) {
        return new TitaniumLiteral.LangLiteral(lex, lang);
    }

    @Override
    public Object makeDtLiteral(String lex, String dt) {
        return new TitaniumLiteral.DtLiteral(lex, dt);
    }

    @Override
    public String makeDatatype(String dt) {
        return dt;
    }

    @Override
    public Object makeBlankNode(String label) {
        return "_:".concat(label);
    }

    @Override
    public Object makeIriNode(String iri) {
        return iri;
    }

    @Override
    public Object makeTripleNode(Object s, Object p, Object o) {
        throw new UnsupportedOperationException(
            "The titanium-rdf-api implementation of Jelly does not support quoted triples."
        );
    }

    @Override
    public Object makeDefaultGraphNode() {
        return null;
    }
}
